ccmlib/node.py

```diff
@@ -560,19 +576,32 @@ def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbos
                             if len(tofind) == 0:
                                 return matchings[0] if isinstance(exprs, string_types) else matchings
                 else:
-                    # yep, it's ugly
+                    # wait for the situation to clarify, either stop or just a pause in log production
                     time.sleep(1)
-                    if start + timeout < time.time():
-                        raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str([e.pattern for e in tofind]) + ":\n" + reads[:50] + ".....\nSee {} for remainder".format(filename))
 
-                if process:
-                    if common.is_win():
-                        if not self.is_running():
-                            return None
-                    else:
-                        process.poll()
-                        if process.returncode == 0:
-                            return None
+                    # check if abort condition is not met e.g. C* process is terminated
+                    if abort_function is not None:
+                        abort_function()
+
+                    # check if timeout passed
+                    check_timeout("Missing: " + str([e.pattern for e in tofind]) +
+                                  ":\n" + reads[:50] +
+                                  ".....\nSee {} for remainder".format(filename))
+
+                    # Checking "process" is tricky, as it may be itself terminated e.g. after "verbose"
+                    # or if there is some race condition between log checking and start process finish
+                    # I am simply afraid to completely remove it because this is general usage method
+                    # so instead we will use it IFF the pid-based abort function is not provided
+                    if process and abort_function is None:
```

Comment:
isn't `abort_function is None` redundant?

Reply:
Not sure why redundant?
It is indeed checked earlier BUT it cannot not throw (it is conditional).
Indeed, I need to change naming ;-)


this is what makes this whole change so complex:
process - is a process that started C*
abort function is a function based e.g. on C* process pid (different one).

Now there are interesting races:
this process can be terminated before C* process is terminated.
in such case instead of finding C* process dead => raise NodeError
this logic will return None and continue
and then it will timeout on the next step or fail silently